File: SfunctionsR/src_user/user_files/user_ExtForces.c
//--------------------------- 
// UCL-CEREM-MBS 
// 
// @version MBsysLab_s 1.7.a 
// 
// Creation : 2006 
// Last update : 01/10/2008 
//--------------------------- 

#include "simu_def.h"

double* user_ExtForces(double PxF[4], double RxF[4][4], 
					   double VxF[4], double OMxF[4], 
					   double AxF[4], double OMPxF[4], 
					   MBSdataStruct *MBSdata, double tsim, int ixF)
{
	int i;
	double Fx=0.0, Fy=0.0, Fz=0.0;
	double Mx=0.0, My=0.0, Mz=0.0;
	double dxF[3+1] ={0.0, 0.0, 0.0, 0.0}; // +1 because indexes begin at 1

	double *SWr = MBSdata->SWr[ixF];
	int idpt = 0;
        
    // user variables
    UserIOStruct *uvs;
    
    uvs = MBSdata->user_IO;

	idpt = MBSdata->xfidpt[ixF];

	dxF[1] = MBSdata->dpt[1][idpt];
	dxF[2] = MBSdata->dpt[2][idpt];
	dxF[3] = MBSdata->dpt[3][idpt];

	// initializes the terms of Swr
	SWr[1] = Fx;
	SWr[2] = Fy;
	SWr[3] = Fz;
	SWr[4] = Mx;
	SWr[5] = My;
	SWr[6] = Mz;
	SWr[7] = dxF[1];
	SWr[8] = dxF[2];
	SWr[9] = dxF[3];

	return SWr;
}

File: userfctR/user_ExtForces.m
function Swr = user_ExtForces(PxF,RxF,VxF,OMxF,AxF,OMPxF,mbs_data,tsim,ixF)
% --------------------------
% UCL-CEREM-MBS
%
% @version MBsysLab_m 1.7.a
%
% Creation : 2006
% Last update : 30/09/2008
% -------------------------
%
%Swr = user_ExtForces(PxF,RxF,VxF,OMxF,AxF,OMPxF,mbs_data,tsim,ixF)
%
% PxF(3,1) : absolute position vector of the external force application point 
% RxF(3,3) : absolute rotation matrix of the body
% VxF(3,1) : absolute velocity vector of the external force application point 
% OMxF(3,1) : absolute angular velocity vector of the body
% AxF(3,1) : absolute acceleration vector of the external force application point 
% OMPxF(3,1) : absolute angular acceleration vector of the body
%
% => All above vectors are expressed in the inertial reference frame !
%
% mbs_data : multibody data structure
% tsim : current time
% ixF : index of the external force sensor ('F' type in MBsysPad)
%        (can be obtained via the 'mbs_get_F_sensor_id' function)
%
% Swr(9,1) = [Fx; Fy; Fz; Mx; My; Mz; dxF];
%   - Force components (expressed in the inertial frame) : Fx, Fy, Fz
%   - Pure torque components (expressed in the inertial frame) : Mx, My, Mz
%   - Application point local coordinates vector (expressed in the body-fixed frame) : dxF(1:3,1);
%
% this function may use a global structure called MBS_user

global MBS_user MBS_info

Fx=0.0; Fy=0.0; Fz=0.0;
Mx=0.0; My=0.0; Mz=0.0;
idpt = mbs_data.xfidpt(ixF);
dxF = mbs_data.dpt(:,idpt);

%/*-- Begin of user code --*/
% 
% Use the 'mbs_get_F_sensor_id' function to get easily the force sensor
% indices, e.g. :
% F1 = mbs_get_F_sensor_id(MBS_info,'myFsensor_1');
% [F2,F3] = mbs_get_F_sensor_id(MBS_info,'myFsensor_2','myFsensor_3');
%

F_RF_Leg = mbs_get_F_sensor_id(MBS_info,'FR_Leg_Force');
F_RL_Leg = mbs_get_F_sensor_id(MBS_info,'FL_Leg_Force');
F_RR_Leg = mbs_get_F_sensor_id(MBS_info,'RR_Leg_Force');
F_LR_Leg = mbs_get_F_sensor_id(MBS_info,'RL_Leg_Force');

switch(ixF)
    case F_RF_Leg
        if(PxF(3) < 0)
            Fz = 10000*abs(PxF(3)) - 100*VxF(3);
            Fx = (-2000.9)*VxF(1);
        end
        
    case F_RL_Leg
        if(PxF(3) < 0)
          Fz = 10000*abs(PxF(3)) - 100*VxF(3);
            Fx = (-2000.9)*VxF(1);
        end
        
    case F_RR_Leg
        if(PxF(3) < 0)
          Fz = 10000*abs(PxF(3)) - 100*VxF(3);
            Fx = (-2000.9)*VxF(1);
        end
        
    case F_LR_Leg
        if(PxF(3) < 0)
           Fz = 10000*abs(PxF(3)) - 100*VxF(3);
            Fx = (-2000.9)*VxF(1);
        end
end

%/*-- End of user code --*/

Swr = [Fx; Fy; Fz; Mx; My; Mz; dxF];

return